### PR TITLE
[ao] Added Quantization QConfig generation to ModelReport class

### PR DESCRIPTION
**Pull request by vspenubarthi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83698
* __->__ #83688

Summary: This adds the capability to generate a QConfigMapping based on
the suggestions of the ModelReport API for the user to use. The only
dependency of this feature is that the callibration is run before the
generation of the QConfigMapping and there is no dependency on the
report generation other than that the observers cannot be removed before
this is called.

Example Usage (after callibration):
```
mapping = mod_report.generate_qconfig_mapping()

prepared_model = quantize_fx.prepare_fx(model, mapping, example_input)

quantized_model = quantize_fx.convert_fx(prepared)
```

This was tested by ensuring that the suggestions generated in the
QConfigMapping are:
1. Correct according to the set backend and data passed through
2. Able to be prepared and converted as a proper config (is a valid
config)

The test for this is a part of the TestFxModelReportClass test suite.

Test Plan: python test/test_quantization.py TestFxModelReportClass.test_qconfig_mapping_generation

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 1]
```diff
@@ -6,9 +6,19 @@
 from abc import ABC, abstractmethod
 from torch.ao.quantization.fake_quantize import FakeQuantize
 from torch.ao.quantization.fx.graph_module import GraphModule
-from torch.ao.quantization.observer import ObserverBase
 from torch.ao.quantization.fx._model_report.model_report_observer import ModelReportObserver
-from torch.ao.quantization.qconfig import QConfig
+from torch.ao.quantization.qconfig import (
+    QConfig,
+    default_qconfig,
+    assert_valid_qconfig,
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/detector.py:13]: please fix the lint

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 2]
```diff
@@ -17,6 +27,66 @@
...
+    def generate_qconfig(self, module: torch.nn.Module) -> QConfig:
+        r"""
+        Args:
+            module (torch.nn.Module) The module we are generating
+            the qconfig for
+
+        Returns the generated QConfig according to what a valid configuration is
+        """
+        # Apply suggestions to new qconfig
+        module_qconfig = default_qconfig
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/detector.py:63]: should the default be `default_qconfig` or None?

Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/detector.py:63]: should be default_qconfig because we want to use some level of quantizatio

**Review by jerryzh168 — APPROVED** [review]
looks great, thanks for decoupling model report and detectors!

File: torch/ao/quantization/fx/_model_report/detector.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/_model_report/detector.py:63]: but using quantization is not always good right? e.g. if detector recommends to use per channel quant based on data, but it's not supported on the operator, then using per tensor quant means it will have poor accuracy, is this something we want to recommend?

Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/detector.py:63]: the detector only recommends to use per channel quantization for scenarios in which both per channel and per tensor are supported by the operator, (it cross checks with list of supported modules)

so we won't run into the case where we recommend per channel quant based on data but it's not supported by operator
